tests/unit/backend/corpora/common/entities/test_dataset.py

```diff
@@ -28,3 +35,91 @@ def test__get__ok(self):
     def test__get__does_not_exist(self):
         non_existent_key = "non_existent_id"
         self.assertEqual(Dataset.get(non_existent_key), None)
+
+    def test__create__ok(self):
+        """
+        Create a dataset with a variable number of artifacts, contributors, and deployment_directories
+
+        """
+        artifact_params = dict(
+            filename="filename_1",
+            filetype=DatasetArtifactFileType.H5AD,
+            type=DatasetArtifactType.ORIGINAL,
+            user_submitted=True,
+            s3_uri="some_uri",
+        )
+
+        deployment_directory_params = dict(environment="test", url="test_url")
+
+        contributor_params = dict(name="bob", institution="school", email="some@email.com")
+
+        dataset_params = dict(
+            name="create_dataset_id",
+            organism="organism",
+            organism_ontology="123",
+            tissue="tissue",
+            tissue_ontology="123",
+            assay="assay",
+            assay_ontology="123",
+            disease="diseas",
+            disease_ontology="123",
+            sex="F",
+            ethnicity="ethnicity",
+            ethnicity_ontology="123",
+            source_data_location="location",
+            preprint_doi="preprint",
+            publication_doi="publication",
+        )
+
+        for i in range(3):
```

Comment:
Curious, why 3? What are the different test cases that are tested by doing it three times?

Reply:
It's the trinity. Everything is three. Also since python range starts at 0, we are effectively testing 0,1, and 2 foreign keys linked to the dataset. Those are usually the critical test areas.

Reply:
Wouldn't you want to separate out a test that tests 0 foreign keys to use the None keyword then? And then a separate test for testing 1?

I don't quite understand the benefit of testing having 1 versus 2 foreign keys. If you have documentation on this testing practice I'd be happy to learn too!

Reply:
It's not so much a test that the DB can manage foreign keys. It's that our code will not fail when parsing 0, 1 or multiple values from the database into python. It's a general QA best practice to test around the edges. Here is an article that encompases the spirit of the tests https://www.freecodecamp.org/news/a-beginners-guide-to-testing-implement-these-quick-checks-to-test-your-code-d50027ad5eed/.

Reply:
Are you not a fan of using subtest?

Reply:
Interesting, thank you for the article. You're right, I'm not a fan of subtests. The reason is because when one of the subtests fail, the failure is somewhat hidden by the fact that the tests are lumped into a single unit test. For example, if I had a test `test__create_dataset_with_no_foreign_keys__ok` that failed versus a `test__create_dataset_with_multiple_foreign_keys__ok` that failed, it immediately tells me where to look rather than having to dive more deeply into the code to figure out which `i` value caused the failure and what that `i` value means and what the `i` value means in terms of the created input to the test.

Reply:
If you look at the output of unittest it tells you exactly which iteration failed. There should be no code diving required. It spits out an error message for each failed iteration as if it was a stand alone test. We can always add a more verbose subtest name. I don't see this as a compelling reason to not use subtests.